### DISCUSSION TITLE
[BE] Service 메서드의 범용성에 대하여

### DISCUSSION THREAD

**Discussion by gitchannn** [woowacourse-teams/2023-stamp-crush#221] [General]
```java
    public Long createCafe(CafeCreateDto cafeCreateDto) {
        Owner owner = findExistingOwnerById(cafeCreateDto.getOwnerId());
        Cafe savedCafe = saveCafe(cafeCreateDto, owner);
        assignDefaultCafePolicyToCafe(savedCafe);
        assignDefaultSampleCafeCouponToCafe(savedCafe);

        return savedCafe.getId();
    }
```

위와 같은 경우에 Service 코드가 id를 반환하는 것은 제 생각에는 `Response`의 형태를 고려해서 만들어진 것 같아요.
이럴 경우에 서비스 코드만을 API로 제공한다고 생각했을 때, `savedCafe`를 그대로 반환하는 것이 더 자연스러울 것 같다고 생각합니다.

```java
 @Test
    void 카페를_등록할_때_쿠폰이미지는_기본이미지로_저장된다() {
        // given
        CafeCreateDto cafeCreateDto = getCafeCreateDto();

        // when
        Long cafeId = cafeService.createCafe(cafeCreateDto);
        Cafe cafe = cafeRepository.findById(cafeId).get(); // 불필요하게 두 번 조회해야함
        Optional<CafeCouponDesign> cafeCouponDesign = cafeCouponDesignRepository.findByCafe(cafe)
                .stream()
                .filter(design -> design.getCafe().getId().equals(cafeId))
                .findAny();

        // then
        SoftAssertions softAssertions = new SoftAssertions();
        softAssertions.assertThat(cafeCouponDesign).isPresent();
        softAssertions.assertThat(cafeCouponDesign.get().getFrontImageUrl()).isEqualTo(sampleFrontImage.getImageUrl());
        softAssertions.assertThat(cafeCouponDesign.get().getBackImageUrl()).isEqualTo(sampleBackImage.getImageUrl());
        softAssertions.assertThat(cafeCouponDesign.get().getStampImageUrl()).isEqualTo(sampleStampImage.getImageUrl());
        softAssertions.assertAll();
    }
```

서비스에서는 Cafe 객체를 그대로 반환하고, Controller에서 cafe.getId()를 통해서 Location을 생성하는 것에 대해서 어떻게 생각하시나요?

---

마찬가지로, 서비스 코드 중에서, Response DTO와 단지 클래스만 분리되었을 뿐, 형태는 동일해서 범용성을 전혀 고려하지 않은 것들이 다수 있는 것 같은데 전부 수정하는 것에 대해서도  어떻게 생각하는지 궁금합니다.

**Comment by jundonghyuk** [comment]
저는 `Cafe` 객체가 서비스 레이어 밖으로 나가는것이 더 위험하다고 생각하는데요 
그래서 엔티티 객체를 그대로 반환하는 건 반대의 입장이에요 `Lazy Loading`으로 인한 프록시객체가 어떻게 동작할지 예상할 수 없으니까여

그리고 조회관련 서비스일 때는 범용성있게 반환해주는건 괜찮은데, 단순 생성이나, 변경, 삭제 등과 같은 서비스 메서드는 그 객체내용을 전부 서비스에서 반환하는 것은 의미가 없다고 생각을합니다. `Location Header`에 쓰일 `Id`를 반환하는 것또한 `Location Header`에 쓰이지 않는다면 이것또한 리턴할 필요가 없다고 생각합니다. 

`범용적`이라는 것은 다른 부분에서도 공통적으로 쓰고 싶다는 의미로 파악했습니다. 그렇지만 어느곳에서도 반환데이터를 쓰지 않는다면 범용적이라고 할 수 없을 것 같습니다. 대댓으로 의견 남겨주시면 유심히! 읽어보겠습니다 !

**Reply by gitchannn** [reply]
이후에 생각을 해봤는데,

명령(save)과 조회를 분리하는 것은 충분히 의미가 있다고 생각이 들어요. 

> 그리고 조회관련 서비스일 때는 범용성있게 반환해주는건 괜찮은데, 단순 생성이나, 변경, 삭제 등과 같은 서비스 메서드는 그 객체내용을 전부 서비스에서 반환하는 것은 의미가 없다고 생각을합니다. Location Header에 쓰일 Id를 반환하는 것또한 Location Header에 쓰이지 않는다면 이것또한 리턴할 필요가 없다고 생각합니다.

이부분도 동의합니다!

내가 범용성이라고 말하려 했던건 `조회 관련 서비스` 코드를 보면서 들었던 생각인 것 같아요.

지난번에 내가 리팩터링 했던 코드중에서 
Response가 List<>를 필드로 갖고 있는 복수형 이름의 Response DTO였는데, 
그걸 고려해서 서비스에서도 한 번 감싼 형태의 복수형 Result DTO를 반환한 적이 있었슴다.
이런식으로 좀 과도하게 리스폰스 형태를 고려한 서비스 반환값을 말한 것이었슴다!

**Comment by youngh0** [comment]
오 저는 레벨2때 `범용적인 DTO를 서비스에서 반환하고 컨트롤러에서 필요한 걸 뽑아서 반환하면 어떨까?` 라는 생각을 해본 적이 있어요. 예를 들면, 서비스는 Coupon의 모든 데이터만 담고 있는 CouponDto만 반환하는 거죠.

리뷰어 한테 이러한 제 의견을 물어봤더니 `범용적인` 이라는 단어에서 SRP를 위반하는 것 같지 않냐는 답변을 받은 적이 있는데요. 생각해보니, 범용적인 객체를 반환하게 되면 해당 객체(범용적인 Dto)는 아주 다양한 이유로 변경될 수도 있을 것 같아요.

이를 바탕으로 생각을 해봤는데, 저는 서비스 메서드 안에서 작은 기능을 수행하는 것은 범용적으로 만들어도, 서비스 메서드 자체를 범용적으로 만드는건 잘 모르겠어요. 
예를 들어, 통계 내는 책임을 Coupons라는 객체에 부여하면 서비스 메서드 내부의 작은 기능을 범용적으로 만든게 되겠죠. 이렇게 되면 또 다른 종류의 통계 데이터를 반환하는 서비스 메서드에서 Coupon의 통계를 낼 때, 이를 사용할 수 있겠죠. 
그러나, 통계 데이터를 반환하는 범용적인 A라는 서비스를 만들었다고 가정해볼게요. 만약, `나이 정보`가 통계에 필요해지게 되면 A서비스를 수정하겠죠. `사는 지역 데이터`가 필요해도 A서비스를 수정하게 될 것 같아요.

제 생각을 예시로 들다보니, 말이 구구절절 길어졌네요.

결론!!!
제목이 `Service메서드의 범용성`이라고 적혀있어서, 모든 컨트롤러가 비슷한 종류의 요청(조회, 생성 등)에서 하나의 서비스 메서드를 사용하게 하자! 라는 의도로 이해했고, 이에 대한 제 생각은 서비스 메서드 자체보다는, 해당 서비스를 완성하기 위한 작은 기능들을 범용성있게 만들어보는게 좋지 않을까?? 입니다.
예시로 들어주신 create의 경우 범용적으로 쓰는 일이 잘 상상이 가지 않아서 혹시 다른 서비스 코드(질문의 마지막 문장)를 예시로 들어주시면 더 깃짱 의도를 더 정확히 파악할 수 있을 것 같습니다~

깃짱의 의도를 정확히 파악하고 댓글은 달았는진 모르겠네요. 제가 잘못 이해한게 있거나 다른 생각을 가지고 계시면 언제든 말해주세요 ^.^ 저 없는동안 고생해주셔서 감사합니다 ㅎㅎ

**Reply by gitchannn** [reply]
흠.. 레오 말을 보고 좀 생각이 생기는군여
나도 좀 생각해보고 다시 말하려고 했던 상황이 보일 때 말하겠슴당

**Comment by gitchannn** [comment]
_size()를 필요로 하는 경우에 Service에서 .size()를 찍어서 정보를 누락하지 말고 컨트롤러에서 .size()를 해서 보내주도록 하자_

**Reply by gitchannn** [reply]
_Service DTO를 리스폰스 따라서 일급컬렉션으로 한 번 더 묶은 것도 불필요할 것 같음!_

**Reply by gitchannn** [reply]
Response DTO = 화면

서비스 DTO와 Response DTO를 정확히 동일한 모양으로 만들면, 두 가지를 분리한 의미가 뭘까?
서비스에서 보내주는 반환값은 조금 더 일반적인 형태로 가는 게 좋을 것 같다.
Response와 Controller는 백엔드에서 화면을 예쁘게 그리지 않다 뿐이지, 화면이라고 볼 수 있다고 생각한다.

**Reply by gitchannn** [reply]
일반적인 형태라는 선이 굉장히 중요한데, 이거는 개발자 개개인의 센스에 달려있을 것 같다.

**Comment by yenawee** [comment]
이 디스커션에 해당하는 상황이 발생한다면 코드 리뷰에 달아주자 땅땅땅
